### DISCUSSION TITLE
Latest Performance

### DISCUSSION THREAD

**Discussion by Marshmallow3535** [pocketbase/pocketbase#2757] [Q&A] [answered]
Hey, I know that you will be/are working on a test bed and benchmarking suite. As it's not yet released I was hoping for some information on aspects of the performance.

1. What is the limiting factor when it comes to vertical scaling, is it CPU, RAM or something else (basically what would I need to upgrade first)

2. How many concurrent real-time users could it handle on a dedicated box with an amd ryzen 5 3600 with 32gb of ram. Obviously I'm just wanting a ballpark guess.

Thankyou in advanced, also thanks for your amazing work on pocketbase so far. I've been using it for one of my latest projects and it's been really intuitive and a joy to use so far.

**Comment by ganigeorgiev** [comment, marked as answer]
Without any context about what is your use case and the expected workload I don't think I can answer the questions properly.

If you hit a performance bottleneck, the easiest way to speed up the db operations without changing your hardware is to build PocketBase with cgo (aka. `CGO_ENABLED=1 go build`). Based on the most recent [cgo vs go sqlite drivers benchmark](https://docs.google.com/spreadsheets/d/1YOP1D_ZhuR-ednQhTH6SSrCkcR5bOHzCDoQBZMXvyAw/edit#gid=9643931) the cgo driver could be ~2x  times faster for some queries. The drawback is that the binary will no longer be statically linked by default but if you are building on Linux (or in a container) it shouldn't be much of an issue.

When hardware upgrade is needed, I'd say prioritize first faster ssd/nvme disk (if you are not using one already) and then based on some monitoring, upgrade the ram and cpu.

My only performance benchmarks so far was with the default go SQLite port against Hetzner CX21 VPS (2vCPUs, 4GB) with GET requests to various sized collections (20k, 50k, 100k) and was able to get to stable ~10k+ concurrent clients (you may have to increase the OS allowed file descriptors limit; see https://pocketbase.io/docs/going-to-production/#increase-the-open-file-descriptors-limit).

Official testbed and benchmarks repo will be available soon (probably with the next release) and you'll be able to test it on your own. Until then, if you manage to create your own benchmarks test feel free to share it.

**Reply by Marshmallow3535** [reply]
Thanks that's really helpful

**Comment by HEAVYPOLY** [comment]
Is using S3 (wasabi) slower vs storing files on pocketbase directly?  Also if selfhosting pocketbase in US, is speed slower the farther away the user is from server?

**Reply by ganigeorgiev** [reply]
> Is using S3 (wasabi) slower vs storing files on pocketbase directly?

S3 will be always slower compared to direct accessing the files from the disk since the files'll need to be fetched over the network using the S3 apis.

> Also if selfhosting pocketbase in US, is speed slower the farther away the user is from server?

Yes, the location of the server can affect the loading speed. This also depends on the host network, but in general you should expect better performance if the server is closer to your users.

**Reply by Marshmallow3535** [reply]
Apologies if this is somewhere in the documentation, but does the SQLite database get stored on S3 or is it just uploaded files and backups? Thanks

**Reply by HEAVYPOLY** [reply]
Thanks, loving pocketbase so far, I'm very new to databases but everything is making sense.  

On this note is hosting your actual site on pocketbase/public faster than having it hosted elsewhere (github pages etc)
Also curious if users in China are able to access servers hosted in the US

**Reply by ganigeorgiev** [reply]
@Marshmallow3535 

> does the SQLite database get stored on S3 or is it just uploaded files and backups? 

No, the SQlite file is stored only on the disk. As you've already noted, in S3 you could store the uploaded files and the backups.

---

@HEAVYPOLY 

> On this note is hosting your actual site on pocketbase/public faster than having it hosted elsewhere (github pages etc)

GitHub pages to some extend act also as a CDN since technically your static content would be served from different locations around the world (and most likely is load balanced) but keep in mind that there are also rate limit rules and other restrictions that may affect your users (see also their ["Usage limits" section](https://docs.github.com/en/pages/getting-started-with-github-pages/about-github-pages#usage-limits)).

> Also curious if users in China are able to access servers hosted in the US

This is difficult to answer but if the sever location is a concern, the best recommendation would be to test your host provider with a VPN service and measure it based on your use case and expected traffic.

**Comment by pelep618** [comment]
Good
